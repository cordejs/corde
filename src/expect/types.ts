declare namespace corde {
  interface IMatchers<T extends any> {
    toBe(expected: T): void;
    toEqual(expected: T): void;
    toBeCloseTo(expected: T, precisionopt: number): void;
    toBeDefined(): void;
    toBeFalse(): void;
    toBeFalsy(): void;
    toBeGreaterThan(expected: number | bigint): void;
    toBeGreaterThanOrEqual(expected: number | bigint): void;
    toBeInstanceOf<U extends any>(expected: T | U): void;
    toBeLessThan(expected: number | bigint): void;
    toBeLessThanOrEqual(expected: number | bigint): void;
    toBeNaN(): void;
    toBeNegativeInfinity(): void;
    toBeNull(): void;
    toBeUndefined(): void;
    toBePositiveInfinity(): void;
    toBeTrue(): void;
    toBeTruthy(): void;
    toContain(expected: T): void;
    toHaveBeenCalled(): void;
    toHaveBeenCalledBefore(expected): void;
    toHaveBeenCalledOnceWith(): void;
    toHaveBeenCalledTimes(amount: number): void;
    toHaveBeenCalledWith(): void;
    toHaveClass(expected): void;
    toHaveSize(size: number): void;
    toMatch(expected): void;
    toThrow<U extends any>(expectedThrow: U): void;
    toThrowError<U extends Error>(expectedopt: U): void;
    toThrowMatching(predicate): void;
    any(constructor: any): void;
    anything(mixed: any): void;
    arrayContaining(mixed): void;
    objectContaining(mixed): void;
    stringMatching(pattern): void;
    toBeArrayOfBooleans(): void;
    toBeArrayOfNumbers(): void;
    toBeArrayOfObjects(): void;
    toBeArrayOfSize(number): void;
    toBeArrayOfStrings(): void;
    toBeEmptyArray(): void;
    toBeNonEmptyArray(): void;
    toBeAfter(otherDate: Date): void;
    toBeBefore(otherDate: Date): void;
    toBeDate(): void;
    toBeValidDate(): void;
    toBeFunction(): void;
    toThrowAnyError(): void;
    toThrowErrorOfType(constructorName): void;
    toBeCalculable(): void;
    toBeEvenNumber(): void;
    toBeGreaterThanOrEqualTo(otherNumber: number | bigint): void;
    toBeLessThanOrEqualTo(otherNumber: number | bigint): void;
    toBeNear(otherNumber: number | bigint, epsilon): void;
    toBeOddNumber(): void;
    toBeWholeNumber(): void;
    toBeWithinRange(floor, ceiling): void;
    toBeEmptyObject(): void;
    toBeNonEmptyObject(): void;
    toHaveArray(memberName): void;
    toHaveArrayOfBooleans(memberName): void;
    toHaveArrayOfNumbers(memberName): void;
    toHaveArrayOfObjects(memberName): void;
    toHaveArrayOfSize(memberName, size): void;
    toHaveArrayOfStrings(memberName): void;
    toHaveBoolean(memberName): void;
    toHaveCalculable(memberName): void;
    toHaveDate(memberName): void;
    toHaveDateAfter(memberName, date): void;
    toHaveDateBefore(memberName, date): void;
    toHaveEmptyArray(memberName): void;
    toHaveEmptyObject(memberName): void;
    toHaveEmptyString(memberName): void;
    toHaveEvenNumber(memberName): void;
    toHaveFalse(memberName): void;
    toHaveHtmlString(memberName): void;
    toHaveIso8601(memberName): void;
    toHaveJsonString(memberName): void;
    toHaveMember(memberName): void;
    toHaveMethod(memberName): void;
    toHaveNonEmptyArray(memberName): void;
    toHaveNonEmptyObject(memberName): void;
    toHaveNonEmptyString(memberName): void;
    toHaveNumber(memberName): void;
    toHaveNumberWithinRange(memberName, floor, ceiling): void;
    toHaveObject(memberName): void;
    toHaveOddNumber(memberName): void;
    toHaveString(memberName): void;
    toHaveStringLongerThan(memberName: keyof T, string): void;
    toHaveStringSameLengthAs(memberName: keyof T, string): void;
    toHaveStringShorterThan(memberName: keyof T, string): void;
    toHaveTrue(memberName: keyof T): void;
    toHaveUndefined(memberName: keyof T): void;
    toHaveWhitespaceString(memberName: keyof T): void;
    toHaveWholeNumber(memberName: keyof T): void;
    toBeRegExp(): void;
    toBeEmptyString(): void;
    toBeHtmlString(): void;
    toBeIso8601(): void;
    toBeJsonString(): void;
    toBeLongerThan(otherString: string): void;
    toBeNonEmptyString(): void;
    toBeSameLengthAs(otherString: string): void;
    toBeShorterThan(otherString: string): void;
    toBeString(): void;
    toBeNumber(): void;
    toBeArray(): void;
    toBeObject(): void;
    toLength(length: number): void;
    toBeBoolean(): void;
    toBeWhitespace(): void;
    toEndWith(substring: string): void;
    toStartWith(substring: string): void;
  }

  interface Expect {
    expect<T extends any>(expect: T): corde.IMatchers<T>;
  }
}
