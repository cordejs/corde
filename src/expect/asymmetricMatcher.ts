import { typeOf } from "../utils";

export class AsymmetricMatcher {
  private _classType: any[];

  constructor(...classType: any[]) {
    this._classType = classType;
  }

  isSpecified() {
    return this._classType.length > 0;
  }

  toString() {
    if (this.isSpecified()) {
      const names = this._classType.map((type) => this.getTypeName(type));
      return `any(${names.join(", ")})`;
    }
    return "any";
  }

  getTypeName(classType: any) {
    if (classType === Number) {
      return "Number";
    }

    if (classType === BigInt) {
      return "Bigint";
    }

    if (classType === String) {
      return "String";
    }

    if (classType === Boolean) {
      return "Boolean";
    }

    if (classType === Array) {
      return "Array";
    }

    if (classType === Object) {
      return "Object";
    }

    if (classType === Function) {
      return "Function";
    }

    if (classType === Symbol) {
      return "Symbol";
    }

    if (typeof classType === "function") {
      return classType.name;
    }

    return classType;
  }

  matchType(classType: any) {
    return (
      this._classType === undefined ||
      this._classType === null ||
      this._classType.length === 0 ||
      this._classType.some((type) => classType === type)
    );
  }

  matchValue(value: any) {
    return this._classType.every((classType) => this._isValueOfType(classType, value));
  }

  private _isValueOfType(classType: any, value: any) {
    if (classType === Number) {
      return typeOf(value) === "number";
    }

    if (classType === BigInt) {
      return typeOf(value) === "bigint";
    }

    if (classType === String) {
      return typeOf(value) === "string";
    }

    if (classType === Boolean) {
      return typeOf(value) === "boolean";
    }

    if (classType === Array) {
      return typeOf(value) === "array";
    }

    if (classType === Object) {
      return typeOf(value) === "object";
    }

    if (classType === Function) {
      return typeOf(value) === "function";
    }

    if (classType === Symbol) {
      return typeOf(value) === "symbol";
    }

    if (typeOf(value) === "object") {
      return value instanceof classType;
    }

    return false;
  }
}

export function any(...classType: any[]) {
  return new AsymmetricMatcher(...classType);
}
