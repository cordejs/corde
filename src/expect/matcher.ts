import { ColorResolvable } from "discord.js";
import { testCollector } from "../common/testCollector";
import {
  IEmoji,
  IMessageIdentifier,
  IMessageEditedIdentifier,
  IMessageEmbed,
  Primitive,
  IRoleIdentifier,
  ITestReport,
  ICordeBot,
  TestFunctionType,
  IMessageMatches,
  IRoleMatches,
} from "../types";
import { Colors } from "../utils/colors";
import { RolePermission } from "../utils/permission";
import {
  ToAddReaction,
  ToDeleteRole,
  ToRenameRole,
  ToReturn,
  ToSetRoleColor,
  ToSetRoleHoist,
  ToSetRoleMentionable,
  ToSetRolePosition,
  ToEditMessage,
  ToPinMessage,
  ToRemoveReaction,
  ToUnPinMessage,
  ToSetRolePermission,
  ToEmbedMatch,
  ToMessageContentContains,
} from "./matches";
import { ExpectTest } from "./matches/expectTest";
import { buildReportMessage, resolveName, stringIsNullOrEmpty } from "../utils";
import { getStackTrace } from "../utils/getStackTrace";
import { runtime } from "../common/runtime";
import { IMacherContructorArgs, MayReturnMatch, IExpectTestBaseParams } from "../types";
import { IToHaveResult } from "./matches/toHaveResult";

class BaseMatcher {
  protected _commandName: unknown;
  protected _isNot: boolean;
  protected _isCascade: boolean;
  protected _guildId?: string;
  protected _channelId?: string;
  protected _channelIdToSendCommand?: string;

  constructor({
    commandName,
    isNot,
    isCascade,
    channelId,
    guildId,
    channelIdToSendCommand,
  }: IMacherContructorArgs) {
    this._commandName = commandName;
    this._isNot = isNot ?? false;
    this._isCascade = isCascade ?? false;
    this._guildId = guildId;
    this._channelId = channelId;
    this._channelIdToSendCommand = channelIdToSendCommand;
  }

  // Trace can not me added inside operationFactory because it do,
  // it will get irrelevant data.

  protected async operationFactory<T extends ExpectTest>(
    trace: string,
    type: new (params: IExpectTestBaseParams) => T,
    cordeBot: ICordeBot,
    ...params: Parameters<T["action"]>
  ): Promise<ITestReport> {
    const commandName = await resolveName(this._commandName);

    const op = new type({
      cordeBot,
      command: commandName,
      isNot: this._isNot,
      timeout: runtime.timeOut,
      isCascade: this._isCascade,
      guildId: this._guildId ?? runtime.guildId,
      channelId: this._channelId ?? runtime.channelId,
      channelIdToSendCommand: this._channelIdToSendCommand,
    });

    if (
      !this._isCascade &&
      (commandName == undefined ||
        (typeof commandName === "string" && stringIsNullOrEmpty(commandName)))
    ) {
      return {
        pass: false,
        message: buildReportMessage("command can not be null or an empty string"),
        testName: op.toString(),
      };
    }

    const report = await op.action(...params);

    if (!report) {
      return {
        pass: false,
        message: buildReportMessage("no report was provided by the test"),
        testName: op.toString(),
        trace,
      };
    }

    if (report.pass) {
      return report;
    }

    report.trace = trace;
    return report;
  }

  protected returnOrAddToCollector(testFunction: TestFunctionType): any {
    if (this._isCascade) {
      return testFunction;
    }

    return testCollector.addTestFunction(testFunction);
  }
}

export class MessageMatches<TReturn extends MayReturnMatch>
  extends BaseMatcher
  implements IMessageMatches<TReturn>
{
  toEmbedMatch(embed: IMessageEmbed): TReturn {
    const trace = getStackTrace(undefined, true, "toEmbedMatch");
    return this.returnOrAddToCollector((cordeBot) =>
      this.operationFactory(trace, ToEmbedMatch, cordeBot, embed),
    );
  }

  toMessageContentContains(expectedContent: string): TReturn {
    const trace = getStackTrace(undefined, true, "toMessageContentContains");
    return this.returnOrAddToCollector((cordeBot) =>
      this.operationFactory(trace, ToMessageContentContains, cordeBot, expectedContent),
    );
  }

  toReturn(expect: Primitive | IMessageEmbed) {
    const trace = getStackTrace(undefined, true, "toReturn");
    return this.returnOrAddToCollector((cordeBot) =>
      this.operationFactory(trace, ToReturn, cordeBot, expect),
    );
  }

  toEditMessage(
    newValue: Primitive | IMessageEmbed,
    messageIdentifier?: IMessageEditedIdentifier | string,
  ) {
    const trace = getStackTrace(undefined, true, "toEditMessage");
    return this.returnOrAddToCollector((cordeBot) => {
      return this.operationFactory(trace, ToEditMessage, cordeBot, newValue, messageIdentifier);
    });
  }

  toPin(messageIdentifier: string | IMessageIdentifier) {
    const trace = getStackTrace(undefined, true, "toPin");
    return this.returnOrAddToCollector((cordeBot) => {
      return this.operationFactory(trace, ToPinMessage, cordeBot, messageIdentifier);
    });
  }

  toUnPin(messageIdentifier: string | IMessageIdentifier) {
    const trace = getStackTrace(undefined, true, "toUnPin");
    return this.returnOrAddToCollector((cordeBot) => {
      return this.operationFactory(trace, ToUnPinMessage, cordeBot, messageIdentifier);
    });
  }

  toAddReaction(
    emojis: string[] | IEmoji[] | (string | IEmoji)[],
    messageIdentifier?: string | IMessageIdentifier,
  ) {
    const trace = getStackTrace(undefined, true, "toAddReaction");
    return this.returnOrAddToCollector((cordeBot) =>
      this.operationFactory(trace, ToAddReaction, cordeBot, emojis, messageIdentifier),
    );
  }

  toRemoveReaction(
    emojis: string[] | IEmoji[] | (string | IEmoji)[],
    messageIdentifier?: string | IMessageIdentifier,
  ) {
    const trace = getStackTrace(undefined, true, "toRemoveReaction");
    return this.returnOrAddToCollector((cordeBot) =>
      this.operationFactory(trace, ToRemoveReaction, cordeBot, emojis, messageIdentifier),
    );
  }
}

export class ToHaveResultMatcher extends BaseMatcher {
  toHaveResult(...tests: TestFunctionType[]): void {
    const trace = getStackTrace(undefined, true, "todoInCascade");
    testCollector.addTestFunction((cordeBot) => {
      // Encapsulate the functions inside another so it do not be executed.
      const testEnhanced = tests.map((test) => () => test(cordeBot));
      return this.operationFactory(trace, IToHaveResult, cordeBot, ...testEnhanced);
    });
  }
}

export class RoleMatchesImpl<TReturn extends MayReturnMatch>
  extends BaseMatcher
  implements IRoleMatches<TReturn>
{
  toRenameRole(newName: string, roleIdentifier: IRoleIdentifier | string) {
    const trace = getStackTrace(undefined, true, "toRenameRole");
    return this.returnOrAddToCollector((cordeBot) => {
      return this.operationFactory(trace, ToRenameRole, cordeBot, newName, roleIdentifier);
    });
  }

  toSetRoleColor(color: ColorResolvable | Colors, roleIdentifier: string | IRoleIdentifier) {
    const trace = getStackTrace(undefined, true, "toSetRoleColor");
    return this.returnOrAddToCollector((cordeBot) => {
      return this.operationFactory(trace, ToSetRoleColor, cordeBot, color, roleIdentifier);
    });
  }

  toDeleteRole(roleIdentifier: string | IRoleIdentifier) {
    const trace = getStackTrace(undefined, true, "toDeleteRole");
    return this.returnOrAddToCollector((cordeBot) => {
      return this.operationFactory(trace, ToDeleteRole, cordeBot, roleIdentifier);
    });
  }

  toSetRoleMentionable(mentionable: boolean, roleIdentifier: string | IRoleIdentifier) {
    const trace = getStackTrace(undefined, true, "toSetRoleMentionable");
    return this.returnOrAddToCollector((cordeBot) => {
      return this.operationFactory(
        trace,
        ToSetRoleMentionable,
        cordeBot,
        mentionable,
        roleIdentifier,
      );
    });
  }

  toSetRoleHoist(hoist: boolean, roleIdentifier: string | IRoleIdentifier) {
    const trace = getStackTrace(undefined, true, "toSetRoleHoist");
    return this.returnOrAddToCollector((cordeBot) => {
      return this.operationFactory(trace, ToSetRoleHoist, cordeBot, hoist, roleIdentifier);
    });
  }

  toSetRolePosition(newPosition: number, roleIdentifier: string | IRoleIdentifier) {
    const trace = getStackTrace(undefined, true, "toSetRolePosition");
    return this.returnOrAddToCollector((cordeBot) => {
      return this.operationFactory(trace, ToSetRolePosition, cordeBot, newPosition, roleIdentifier);
    });
  }

  toSetRolePermission(roleIdentifier: string | IRoleIdentifier, ...permissions: RolePermission[]) {
    const trace = getStackTrace(undefined, true, "toSetRolePermission");
    return this.returnOrAddToCollector((cordeBot) => {
      return this.operationFactory(
        trace,
        ToSetRolePermission,
        cordeBot,
        roleIdentifier,
        permissions,
      );
    });
  }
}
