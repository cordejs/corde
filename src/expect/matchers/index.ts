export * from "./toBeArray";
export * from "./toBeBigint";
