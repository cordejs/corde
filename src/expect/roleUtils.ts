import { Role } from "discord.js";
import { IRoleIdentifier } from "../types";
import { formatObject } from "../utils";

class RoleUtils {
  createExpectedMessageForRoleData(roleIdentifier: IRoleIdentifier) {
    if (!roleIdentifier) {
      return null;
    }

    if (roleIdentifier.id && roleIdentifier.name) {
      return `role with id ${roleIdentifier.id} or name '${roleIdentifier.name}'`;
    }

    if (roleIdentifier?.id) {
      return `role with id ${roleIdentifier.id}`;
    }

    if (roleIdentifier?.name) {
      return `role with name '${roleIdentifier.name}'`;
    }

    return null;
  }

  getErrorForUndefinedRoleData(roleIdentifier: IRoleIdentifier) {
    if (!roleIdentifier) {
      return "expected: data to identifier the role (id or name)\n" + "received: null";
    }

    return null;
  }

  validateRole(role: Role | undefined, roleIdentifier: IRoleIdentifier): string | undefined {
    if (!role) {
      const message = roleUtils.createExpectedMessageForRoleData(roleIdentifier);

      if (message) {
        return `expected: ${message}\n` + "received: null";
      }

      return (
        "expected: a id or a name to identify the role\n" +
        `received: ${formatObject(roleIdentifier)}`
      );
    }

    return undefined;
  }

  getRoleData(roleIdentifier: string | IRoleIdentifier) {
    let data: IRoleIdentifier;
    if (typeof roleIdentifier === "string") {
      data = { id: roleIdentifier };
    } else {
      data = roleIdentifier as IRoleIdentifier;
    }
    return data;
  }
}

const roleUtils = new RoleUtils();
export { roleUtils };
