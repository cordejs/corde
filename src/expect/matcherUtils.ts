import { AsymmetricMatcher } from "./asymmetricMatcher";

export namespace matcherUtils {
  export function match(assertFn: () => boolean, expected: any, anyType?: any) {
    if (expected instanceof AsymmetricMatcher) {
      return expected.matchType(anyType);
    } else {
      return assertFn();
    }
  }
}
