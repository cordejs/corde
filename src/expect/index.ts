/* eslint-disable no-console */
import chalk from "chalk";
import { runtime } from "../common/runtime";
import { ITestProps, ITestReport } from "../types";
import { corde } from "../types/globals";
import { getStackTrace, typeOf } from "../utils";
import { any } from "./asymmetricMatcher";
import * as matchers from "./matchers";

interface IMatcher {
  (props: ITestProps, ...args: any[]): ITestReport;
}

type KeyOfMatcher = keyof typeof matchers;

function pickFn(name: KeyOfMatcher) {
  return matchers[name] as any as IMatcher;
}

function createMatcherFn(matcher: string, isNot: boolean, expected: any, isDebug: boolean) {
  return (...args: any[]): ITestReport | void => {
    // If someone pass expect.any, we must invoke it to return
    // the Any matcher.

    if (expected === any) {
      args = [expected(), ...args];
    } else {
      args = [expected, ...args];
    }

    args = args.map((arg) => {
      if (arg === any) {
        return arg();
      }
      return arg;
    });

    const trace = getStackTrace(undefined, true, matcher);
    try {
      const matcherFn = pickFn(matcher as KeyOfMatcher);

      const props: ITestProps = {
        isNot,
        expectedColorFn: chalk.bold,
        receivedColorFn: chalk.bold,
        createHint: (...paramsName: string[]) => {
          return `expect(${chalk.green("expected")}).${isNot ? "not." : ""}${matcher}(${paramsName
            .map((param) => chalk.red(param))
            .join(",")})`;
        },
        formatValue: (value: any) => {
          if (typeof value === "symbol") {
            return value.toString();
          }

          if (typeOf(value) === "object") {
            if (Object.keys(value).length === 0) {
              return "{}";
            }

            return "{ ... }";
          }

          if (
            typeof value === "string" &&
            (value.length === 0 || (!value.startsWith("'") && !value.endsWith("'")))
          ) {
            return `'${value}'`;
          }

          if (Array.isArray(value)) {
            if (value.length > 0) {
              return "[...]";
            }
            return "[]";
          }

          return value;
        },
      };

      const report = matcherFn.bind(props, ...args)();
      if (!report.pass) {
        report.trace = trace;
      }

      runtime.internalEvents.emit("test_end", report);

      if (isDebug) {
        return report;
      }
    } catch (error) {
      const failedReport: ITestReport = {
        pass: false,
        message: handleError(error),
        trace: trace,
      };
      runtime.internalEvents.emit("test_end", failedReport);
      if (isDebug) {
        return failedReport;
      }
    }
  };
}

function handleError(error: any) {
  if (error instanceof Error) {
    return error.message;
  }
  return error;
}

function createLocalExpect(isDebug: boolean) {
  let localExpect: any = {};

  localExpect = (expected: any) => {
    const _expect: any = {};
    _expect.not = {};
    Object.getOwnPropertyNames(matchers).forEach((matcher) => {
      _expect[matcher] = createMatcherFn(matcher, false, expected, isDebug);
      _expect.not[matcher] = createMatcherFn(matcher, true, expected, isDebug);
    });
    return _expect;
  };

  localExpect.any = any;

  return localExpect;
}

export const expect = createLocalExpect(false) as corde.IExpect;

type DebugExpectType<T> = {
  [P in keyof T]: T[P] extends (...args: any[]) => any
    ? (...params: Parameters<T[P]>) => { pass: boolean; message: string }
    : DebugExpectType<T[P]>;
};

export interface IDebugExpect {
  <T extends any>(value: T): DebugExpectType<corde.IMatchersWithNot<T>>;
  any(...classType: any[]): any;
}

/**
 * Testing object for corde
 * This is not used in production
 *
 * @internal
 */
export const cordeExpect = createLocalExpect(true) as IDebugExpect;
