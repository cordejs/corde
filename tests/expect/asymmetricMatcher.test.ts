import { any } from "../../src/expect/asymmetricMatcher";

class Class {}

describe("testing any matcher", () => {
  it.each([[Number], [BigInt], [Boolean], [Function], [Object], [Symbol], [Class]])(
    "should return true for any type %s",
    (classType) => {
      expect(any().matchType(classType)).toBeTruthy();
    },
  );

  it.each([[Number], [BigInt], [Boolean], [Function], [Object], [Symbol], [Class]])(
    "should assert type %s",
    (classType) => {
      expect(any(classType).matchType(classType)).toBeTruthy();
    },
  );

  it("should assert with multiples type", () => {
    expect(any(Number, String).matchType(Number)).toBeTruthy();
  });

  it("should return name of asymetric", () => {
    expect(any().toString()).toEqual("any");
  });

  it("should return name of asymetric", () => {
    expect(
      any(Number, BigInt, Boolean, Function, Object, Symbol, Class).toString(),
    ).toMatchSnapshot();
  });

  it("should return false for asymetric declared with no type", () => {
    expect(any().isSpecified()).toBeFalsy();
  });

  it("should return true for asymetric declared with no type", () => {
    expect(any(Number).isSpecified()).toBeTruthy();
  });

  it.each([
    [undefined, undefined],
    [null, null],
  ])("should assert true for type %s, with value %s", (classType, value) => {
    expect(any(classType).matchValue(value)).toBeTruthy();
  });

  it.each([
    [Number, 1],
    [Number, new Number()],
    [String, new String()],
    [Boolean, new Boolean()],
    [Function, new Function()],
    [BigInt, 9007199254740991n],
    [Boolean, false],
    [Function, () => {}],
    [Object, {}],
    [Array, []],
    [Symbol, Symbol.for("")],
    [Class, new Class()],
  ])("should assert type %s, with value %s", (classType, value) => {
    expect(any(classType).matchValue(value)).toBeTruthy();
  });
});
