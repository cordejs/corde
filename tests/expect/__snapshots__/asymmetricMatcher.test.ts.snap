// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing any matcher should return name of asymetric 1`] = `"any(Number, Bigint, Boolean, Function, Object, Symbol, Class)"`;
